reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala

```diff
@@ -74,84 +83,132 @@ object StreamSubscriber {
     def onFinalize: F[Unit]
 
     /** producer for downstream */
-    def dequeue1: F[Either[Throwable, Option[A]]]
+    def dequeueChunk1: F[Chunk[Either[Throwable, Option[A]]]]
 
     /** downstream stream */
     def stream(subscribe: F[Unit])(implicit ev: ApplicativeError[F, Throwable]): Stream[F, A] =
       Stream.bracket(subscribe)(_ => onFinalize) >> Stream
-        .eval(dequeue1)
+        .evalUnChunk(dequeueChunk1)
         .repeat
         .rethrow
         .unNoneTerminate
   }
 
-  private[reactivestreams] def fsm[F[_], A](implicit F: Concurrent[F]): F[FSM[F, A]] = {
-    type Out = Either[Throwable, Option[A]]
+  private[reactivestreams] def fsm[F[_], A](q: Queue[F, Either[Throwable, Option[A]]], maxDemand: Int)(
+    implicit F: Concurrent[F]): F[FSM[F, A]] = {
 
     sealed trait Input
     case class OnSubscribe(s: Subscription) extends Input
     case class OnNext(a: A) extends Input
     case class OnError(e: Throwable) extends Input
     case object OnComplete extends Input
     case object OnFinalize extends Input
-    case class OnDequeue(response: Deferred[F, Out]) extends Input
+    case object OnDequeue extends Input
 
     sealed trait State
     case object Uninitialized extends State
     case class Idle(sub: Subscription) extends State
-    case class RequestBeforeSubscription(req: Deferred[F, Out]) extends State
-    case class WaitingOnUpstream(sub: Subscription, elementRequest: Deferred[F, Out]) extends State
+    case class Receiving(sub: Subscription) extends State
+    case object RequestBeforeSubscription extends State
+    case class WaitingOnUpstream(sub: Subscription) extends State
     case object UpstreamCompletion extends State
     case object DownstreamCancellation extends State
     case class UpstreamError(err: Throwable) extends State
 
     def step(in: Input): State => (State, F[Unit]) =
       in match {
         case OnSubscribe(s) => {
-          case RequestBeforeSubscription(req) => WaitingOnUpstream(s, req) -> F.delay(s.request(1))
-          case Uninitialized                  => Idle(s) -> F.unit
+          case RequestBeforeSubscription =>
+            WaitingOnUpstream(s) -> F.delay(s.request(maxDemand))
+          case Uninitialized =>
+            Idle(s) -> F.unit
+          case Idle(_) =>
+            Idle(s) -> F.delay(s.cancel)
```

Comment:
I'm still not convinced this is correct, if we hit this case and then attempt `OnDequeue`, we'll end up requesting demand from the canceled subscription (which will never produce anything) and enter `WaitingOnUpstream`, deadlocking.

Also I think we could handle the `Receiving` state similar to `Idle` instead of erroring.